scripts/gloo2.py

```diff
@@ -0,0 +1,475 @@
+        """ Set a texture sampler. Value is the id of the texture to link.
+        """
+        if not self._linked:
+            raise RuntimeError('Cannot set uniform when program has no code')
+        # Get handle for the uniform, first try cache
+        handle = self._handles.get(name, -1)
+        if handle < 0:
+            if name in self._known_invalid:
+                return
+            handle = self._gl.getUniformLocation(self._handle, name)
+            if name in self._unset_variables:
+                self._unset_variables.remove(name)  # Mark as set
+            self._handles[name] = handle  # Store in cache
+            if handle < 0:
+                self._known_invalid.append(name)
+                print('Variable %s is not an active uniform' % name)
+                return
+        # Program needs to be active in order to set uniforms
+        self.activate()
+        if True:
+            # Sampler: the value is the id of the texture
+            unit = len(self._samplers.keys())
+            if name in self._samplers:
+                unit = self._samplers[name][-1]  # Use existing unit            
+            self._samplers[name] = value._target, value._handle, unit
+            self._gl.uniform1i(handle, unit)
+    
+    def set_uniform(self, name, type_, value):
+        """ Set a uniform value. Value is assumed to have been checked.
+        """
+        if not self._linked:
+            raise RuntimeError('Cannot set uniform when program has no code')
+        # Get handle for the uniform, first try cache
+        handle = self._handles.get(name, -1)
+        count = 1
+        if handle < 0:
+            if name in self._known_invalid:
+                return
+            handle = self._gl.getUniformLocation(self._handle, name)
+            if name in self._unset_variables:
+                self._unset_variables.remove(name)
+            # if we set a uniform_array, mark all as set
+            if not type_.startswith('mat'):
+                count = value.length // (self.ATYPEINFO[type_][0])
+            if count > 1:
+                for ii in range(count):
+                    if '%s[%s]' % (name, ii) in self._unset_variables:
+                        name_ = '%s[%s]' % (name, ii)
+                        if name_ in self._unset_variables:
+                            self._unset_variables.remove(name_)
+            self._handles[name] = handle  # Store in cache
+            if handle < 0:
+                self._known_invalid.add(name)
+                logger.info('Variable %s is not an active uniform' % name)
+                return
+        # Look up function to call
+        funcname = self.UTYPEMAP[type_]
+        # Program needs to be active in order to set uniforms
+        self.activate()
+        self._gl[funcname](handle, value)
+    
+    # todo: Better api
+    def set_attribute(self, name, type_, value):
+        """ Set an attribute value. Value is assumed to have been checked.
```

Comment:
I wrote that todo some time ago. Not sure what I had in mind. Maybe better docs should be sufficient. Talking of docs, what is a sensible way to document the gloo API. Should we put it on RTD?